README
= iCalendar -- Internet calendaring, Ruby style

This is a Ruby library for dealing with iCalendar files.  Rather than
explaining myself, here is the introduction from RFC-2445, which
defines the format:

The use of calendaring and scheduling has grown considerably in the
last decade. Enterprise and inter-enterprise business has become
dependent on rapid scheduling of events and actions using this
information technology. However, the longer term growth of calendaring
and scheduling, is currently limited by the lack of Internet standards
for the message content types that are central to these knowledgeware
applications. This memo is intended to progress the level of
interoperability possible between dissimilar calendaring and
scheduling applications. This memo defines a MIME content type for
exchanging electronic calendaring and scheduling information. The
Internet Calendaring and Scheduling Core Object Specification, or
iCalendar, allows for the capture and exchange of information normally
stored within a calendaring and scheduling application; such as a
Personal Information Manager (PIM) or a Group Scheduling product. 

The iCalendar format is suitable as an exchange format between
applications or systems. The format is defined in terms of a MIME
content type. This will enable the object to be exchanged using
several transports, including but not limited to SMTP, HTTP, a file
system, desktop interactive protocols such as the use of a memory-
based clipboard or drag/drop interactions, point-to-point asynchronous
communication, wired-network transport, or some form of unwired
transport such as infrared might also be used.

Now for some examples:

## Probably want to start with this

  require 'rubygems' # Unless you install from the tarball or zip.
  require 'icalendar'
  require 'date'

  include Icalendar # Probably do this in your class to limit namespace overlap

## Creating calendars and events is easy.

  # Create a calendar with an event (standard method)
  cal = Calendar.new
  cal.event do
    dtstart       Date.new(2005, 04, 29)
    dtend         Date.new(2005, 04, 28)
    summary     "Meeting with the man."
    description "Have a long lunch meeting and decide nothing..."
    klass       "PRIVATE"
  end

  cal.publish

## Or you can make events like this
  event = Event.new
  event.start = DateTime.civil(2006, 6, 23, 8, 30)
  event.summary = "A great event!"
  cal.add_event(event)

  event2 = cal.event  # This automatically adds the event to the calendar
  event2.start = DateTime.civil(2006, 6, 24, 8, 30)
  event2.summary = "Another great event!"

  # Now with support for property parameters
  params = {"ALTREP" =>['"http://my.language.net"'], "LANGUAGE" => ["SPANISH"]}

  cal.event do
    dtstart Date.new(2005, 04, 29)
    dtend   Date.new(2005, 04, 28)
    summary "This is a summary with params.", params
  end
  
  # We can output the calendar as a string to write to a file, 
  # network port, database etc.
  cal_string = cal.to_ical
  puts cal_string

## ALARMS

## Within an event, you can create e-mail notification alarms like this...

  cal.event.do
    # ...other event properties
    alarm do
      action        "EMAIL"
      description   "This is an event reminder" # email body (required)
      summary       "Alarm notification"        # email subject (required)
      attendees     %w(mailto:me@my-domain.com mailto:me-too@my-domain.com) # one or more email recipients (required)
      add_attendee  "mailto:me-three@my-domain.com"
      remove_attendee "mailto:me@my-domain.com"
      trigger       "-PT15M" # 15 minutes before
      add_attach    "ftp://host.com/novo-procs/felizano.exe", {"FMTTYPE" => "application/binary"} # email attachments (optional)
    end
    
    alarm do
      action        "DISPLAY" # This line isn't necessary, it's the default
      summary       "Alarm notification"
      trigger       "-P1DT0H0M0S" # 1 day before
    end

    alarm do
      action        "AUDIO"
      trigger       "-PT15M"
      add_attach    "Basso", {"VALUE" => ["URI"]}  # only one attach allowed (optional)
    end
  end

  # Output

  # BEGIN:VALARM
  # ACTION:EMAIL
  # ATTACH;FMTTYPE=application/binary:ftp://host.com/novo-procs/felizano.exe
  # TRIGGER:-PT15M
  # SUMMARY:Alarm notification
  # DESCRIPTION:This is an event reminder
  # ATTENDEE:mailto:me-too@my-domain.com
  # ATTENDEE:mailto:me-three@my-domain.com
  # END:VALARM
  #
  # BEGIN:VALARM
  # ACTION:DISPLAY
  # TRIGGER:-P1DT0H0M0S
  # SUMMARY:Alarm notification
  # END:VALARM
  # 
  # BEGIN:VALARM
  # ACTION:AUDIO
  # ATTACH;VALUE=URI:Basso
  # TRIGGER:-PT15M
  # END:VALARM

## Timezones
  
  # Create a timezone definition (previous convention)
  cal = Calendar.new
  timezone = Icalendar::Timezone.new
  daylight = Icalendar::Daylight.new
  standard = Icalendar::Standard.new

  timezone.timezone_id =            "America/Chicago"

  daylight.timezone_offset_from =   "-0600"
  daylight.timezone_offset_to =     "-0500"
  daylight.timezone_name =          "CDT"
  daylight.dtstart =                "19700308TO20000"
  daylight.recurrence_rules =       ["FREQ=YEARLY;BYMONTH=3;BYDAY=2SU"]

  standard.timezone_offset_from =   "-0500"
  standard.timezone_offset_to =     "-0600"
  standard.timezone_name =          "CST"
  standard.dtstart =                "19701101T020000"
  standard.recurrence_rules =       ["YEARLY;BYMONTH=11;BYDAY=1SU"]

  timezone.add(daylight)
  timezone.add(standard)
  cal.add(timezone)

  # Now, you can make timezones like this
  cal = Calendar.new
  cal.timezone do
    timezone_id             "America/Chicago"

    daylight do
      timezone_offset_from  "-0600"
      timezone_offset_to    "-0500"
      timezone_name         "CDT"
      dtstart               "19700308TO20000"
      add_recurrence_rule   "FREQ=YEARLY;BYMONTH=3;BYDAY=2SU"
    end

    standard do
      timezone_offset_from  "-0500"
      timezone_offset_to    "-0600"
      timezone_name         "CST"
      dtstart               "19701101T020000"
      add_recurrence_rule   "YEARLY;BYMONTH=11;BYDAY=1SU"
    end
  end
  
  # Both conventions output
  
  # BEGIN:VTIMEZONE
  # TZID:America/Chicago
  # BEGIN:DAYLIGHT
  # TZOFFSETFROM:-0600
  # TZOFFSETTO:-0500
  # TZNAME:CDT
  # DTSTART:19700308T020000
  # RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
  # END:DAYLIGHT
  # BEGIN:STANDARD
  # TZOFFSETFROM:-0500
  # TZOFFSETTO:-0600
  # TZNAME:CST
  # DTSTART:19701101T020000
  # RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
  # END:STANDARD
  # END:VTIMEZONE

== Unicode
Add `$kcode = 'UTF8'` to make icalender work correctly with Utf8 texts
  
== Parsing iCalendars:
  
  # Open a file or pass a string to the parser
  cal_file = File.open("single_event.ics")
  
  # Parser returns an array of calendars because a single file
  # can have multiple calendars.
  cals = Icalendar.parse(cal_file)
  cal = cals.first
  
  # Now you can access the cal object in just the same way I created it
  event = cal.events.first
  
  puts "start date-time: " + event.dtstart
  puts "summary: " + event.summary

== Finders:

Often times in web apps and other interactive applications you'll need to
lookup items in a calendar to make changes or get details.  Now you can find
everything by the unique id automatically associated with all components.
  
  cal = Calendar.new
  10.times { cal.event } # Create 10 events with only default data.
  some_event = cal.events[5] # Grab it from the array of events

  # Use the uid as the key in your app
  key = some_event.uid

  # so later you can find it.
  same_event = cal.find_event(key)

== Examples:

Check the unit tests for examples of most things you'll want to do, but please
send me example code or let me know what's missing.

== Download

The latest release version of this library can be found at

* http://rubyforge.org/projects/icalendar/

Documentation can be found at

* http://icalendar.rubyforge.org/

== Installation

It's all about rubygems:

$ sudo gem install icalendar

== License

This library is released under the same license as Ruby itself.

== Support & Contributions

The iCalendar library homepage is http://icalendar.rubyforge.org/ 

There is an icalendar-devel@rubyforge.org mailing list that can be
used for asking questions, making comments or submitting patches.


History.txt
=== Unreleased

=== 2.10.1 2023-12-01
 * Parsing now handles VTIMEZONE blocks defined after their TZID is used in events and other components

=== 2.10.0 2023-11-01
 * Add changelog metadata to gemspec - Juri Hahn
 * Attempt to rescue timezone info when given a nonstandard tzid with no VTIMEZONE
 * Move Values classes that shouldn't be directly used into Helpers module

=== 2.9.0 2023-08-11
 * Always include the VALUE of Event URLs for improved compatibility - Sean Kelley
 * Improved parse performance - Thomas Cannon
 * Add helper methods for other Calendar method verbs
 * bugfix: Require stringio before use in Parser - vwyu
 
=== 2.8.0 2022-07-10
 * Fix compatibility with ActiveSupport 7 - Pat Allan
 * Set default action of "DISPLAY" on alarms - Rikson
 * Add license information to gemspec - Robert Reiz
 * Support RFC7986 properties - Daniele Frisanco

=== 2.7.1 2021-03-14
 * Recover from bad line-wrapping code that splits in the middle of Unicode code points
 * Add a verbose option to the Parser to quiet some of the chattier log entries

=== 2.7.0 2020-09-12
 * Handle custom component names, with and without X- prefix
 * Fix Component lookup to avoid namespace collisions

=== 2.6.1 2019-12-07
 * Improve performance when generating large ICS files - Alex Balhatchet

=== 2.6.0 2019-11-26
 * Improve performance for calculating timezone offsets - Justin Howard
 * Make it possible to de/serialize with Marshal - Pawel Niewiadomski
 * Avoid FrozenError when running with frozen_string_literal
 * Update minimum Ruby version to supported versions

=== 2.5.3 2019-03-04
 * Improve parsing performance - nehresma
 * Support tzinfo 2.0 - Misty De Meo

=== 2.5.2 2018-12-08
 * Remove usage of the global TimezoneStore instance, in favor of a local variable in the parser
 * Deprecate TimezoneStore class methods

=== 2.5.1 2018-10-30
 * Fix usage without ActiveSupport installed.

=== 2.5.0 2018-09-10
 * Set timezone information from VTIMEZONE components in cases that ActiveSupport can't figure it out (or isn't installed)
 * Prevent rewinding the Parser IO input during parsing - Niels Laukens
 * Update tested/supported ruby versions & documentation updates.

=== 2.4.1 2016-09-03
 * Fix parsing multiple calendars or components in same file - Patrick Schnetger
 * Fix multi-byte folding bug - Niels Laukens
 * Fix typos across the code - Martin Edenhofer & yuuji.yaginuma

=== 2.4.0 2016-07-04
 * Enable parsing individual ICalendar components - Patrick Schnetger
 * many bug fixes. Thanks to Quan Sun, Garry Shutler, Ryan Bigg, Patrick Schnetger and others
 * README/documentation updates. Thanks to JonMidhir and Hendrik Sollich

=== 2.3.0 2015-04-26
 * fix value parameter for properties with multiple values
 * fix error when assigning Icalendar::Values::Array to a component
 * Fall back to Icalendar::Values::Date if Icalendar::Values::DateTime is not given a properly formatted value
 * Downcase the keys in ical_params to ensure we aren't assigning both tzid and TZID

=== 2.2.2 2014-12-27
 * add a `has_#{component}?` method for testing if component exists - John Hope
 * add documentation & tests for organizer attribute - Ben Walding

=== 2.2.1 2014-12-03
 * Prevent crashes when using ActiveSupport::TimeWithZone in multi-property DateTime fields - Danny (tdg5)
 * Ensure TimeWithZone is loaded before using, not just ActiveSupport - Jeremy Evans
 * Improve error message on unparseable DateTimes - Garry Shutler

=== 2.2.0 2014-09-23
 * Default to non-strict parsing
 * Enable sorting events by dtstart - Mark Rickert
 * Better tolerate malformed lines in parser - Garry Shutler
 * Deduplicate timezone code - Jan Vlnas
 * Eliminate warnings - rochefort

=== 2.1.2 2014-09-10
 * Fix timezone abbreviation generation - Jan Vlnas
 * Fix timezone repeat rules for end of month

=== 2.1.1 2014-07-23
 * Quiet TimeWithZone support logging
 * Use SecureRandom.uuid - antoinelyset

=== 2.1.0 2014-06-17
 * Enable parsing all custom properties, not just X- prefixed ones
   Requires non-strict parsing
 * Fixed bugs when using non-MRI ruby interpreters
 * Fix bug copying OpenStruct-backed value types

=== 2.0.1 2014-04-27
 * Re-add support for ruby 1.9.2

=== 2.0.0 2014-04-22
 * Add Icalendar.logger class & logging in Parser
 * Support tzinfo ~> 0.3 and ~> 1.1

=== 2.0.0.beta.2 2014-04-11
 * Add uid & acknowledged fields from valarm extensions
 * Swallow NoMethodError on non-strict parsers
 * Expose a parse_property method on Icalendar::Parser

=== 2.0.0.beta.1 2014-03-30
 * Rewrite for easier development going forward.

=== 1.5.2 2014-02-22
 * Output timezone components first
 * Fix undefined local variable or method 'e' - Jason Stirk

=== 1.5.1 2014-02-27
 * Check for dtend existence before setting timezone - Jonas Grau
 * Clean up and refactor components - Kasper Timm Hansen

=== 1.5.0 2013-12-06
 * Support for custom x- properties - Jake Craige

=== 1.4.5 2013-11-14
 * Fix Geo accessor methods - bouzuya
 * Add ical_multiline_property :related_to for Alarm
 * Allow using multi setters to append single values

=== 1.4.4 2013-11-05
 * Allow user to handle TZInfo::AmbiguousTime error - David Bradford
 * Better handling of multiple exdate and rdate values

=== 1.4.3 2013-09-18
 * Fix concatenation of multiple BYWEEK or BYMONTH recurrence rules

=== 1.4.2 2013-09-11
 * Double Quote parameter values that contain forbidden characters
 * Update Component#respond_to? to match Ruby 2.0 - Keith Marcum

=== 1.4.1 2013-06-25
 * Don't escape semicolon in GEO property - temirov
 * Allow access to various parts of RRule class

=== 1.4.0 2013-05-21
 * Implement ACKNOWLEDGED property for VALARM - tsuzuki08
 * Output VERSION property as first line after BEGIN:VCALENDAR
 * Check for unbounded timezone transitions in tzinfo

=== 1.3.0 2013-03-31
 * Lenient parsing mode ignores unknown properties - David Grandinetti
 * VTIMEZONE positive offsets properly have "+" prepended (Fixed issue
#18) - Benjamin Jorgensen (sorry for misspelling your last name)


examples/create_cal.rb
#!/usr/bin/env ruby
## Need this so we can require the library from the samples directory
$:.unshift(File.dirname(__FILE__) + '/../lib')

require 'rubygems' # Unless you install from the tarball or zip.
require 'icalendar'
require 'date'

include Icalendar # Probably do this in your class to limit namespace overlap

## Creating calendars and events is easy.

# Create a calendar with an event (standard method)
cal = Calendar.new
cal.event do
  dtstart       Date.new(2005, 04, 29)
  dtend         Date.new(2005, 04, 28)
  summary     "Meeting with the man."
  description "Have a long lunch meeting and decide nothing..."
  klass       "PRIVATE"
end

## Or you can make events like this
event = Event.new
event.start = DateTime.civil(2006, 6, 23, 8, 30)
event.summary = "A great event!"
cal.add_event(event)

event2 = cal.event  # This automatically adds the event to the calendar
event2.start = DateTime.civil(2006, 6, 24, 8, 30)
event2.summary = "Another great event!"

# Now with support for property parameters
params = {"ALTREP" =>['"http://my.language.net"'], "LANGUAGE" => ["SPANISH"]} 

cal.event do
  dtstart Date.new(2005, 04, 29)
  dtend   Date.new(2005, 04, 28)
  summary "This is a summary with params.", params
end

# We can output the calendar as a string to write to a file, 
# network port, database etc.
cal_string = cal.to_ical
puts cal_string
